src/database/sqlPersistenceBase.ts

```diff
@@ -0,0 +1,108 @@
+import { knex } from 'knex';
+import { EventLog, PersistenceObject } from '../types';
+import { FilterOperators, FilterTypes, SortClosure, WhereClosure } from './filters';
+
+export abstract class SqlPersistenceBase implements PersistenceObject {
+  protected readonly _knexClient: knex.Knex;
+  
+  constructor(client: 'pg'|'sqlite3') {
+    this._knexClient = knex({ client });
+  }
+
+  abstract init(): Promise<void>;
+
+  abstract saveBatch(batch: EventLog[], blockNumber?: bigint | undefined): Promise<void>;
+
+  abstract getLatestIndexedBlockForChain(chainId: number): Promise<number | undefined>;
+
+  protected abstract getJsonObjectPropertySqlFragment(column: string, propertyName: string): string;
+
+  protected abstract queryAll<T>(query: string): Promise<T[]>;
+
+  abstract disconnect(): Promise<void>;
+
+  public async filter<T extends {}>(
+    table: string,
+    whereClosures: WhereClosure[] = [],
+    sortClosures: SortClosure[] = [],
+    limit: number = 100,
+    offset: number = 0,
+  ): Promise<T[]> {
+    let query = this._knexClient.select('*').from<T>(table);
+
+    for (const c of whereClosures) {
+        c.field.includes('args')
+          ? query.where(
+              this._knexClient.raw(
+                `${this.getJsonObjectPropertySqlFragment(
+                  'args',
+                  c.field.slice(5),
+                )} ${this.getSqlOperator(c.operator)} ?`,
+                [this.convert(c.value, c.type)],
+              ),
+            )
```

Comment:
Knex has a helper for json extract: https://knexjs.org/guide/query-builder.html#jsonextract
I was writing this `getJsonObjectPropertySqlFragment` as an abstraction but if we want to introduce a query builder like knex I'd like it to handle implementation details like this under the hood

Reply:
yes, I totally agree. but as mentioned in the PR description, want to work on this separately from this PR